chronos/futures.nim

```diff
@@ -0,0 +1,369 @@
+#
+#                     Chronos
+#
+#  (c) Copyright 2015 Dominik Picheta
+#  (c) Copyright 2018-2023 Status Research & Development GmbH
+#
+#                Licensed under either of
+#    Apache License, version 2.0, (LICENSE-APACHEv2)
+#                MIT license (LICENSE-MIT)
+
+import std/[strutils, sequtils]
+import stew/base10
+import "."/[config, srcloc]
+export srcloc
+
+when chronosStackTrace:
+  when defined(nimHasStacktracesModule):
+    import system/stacktraces
+  else:
+    const
+      reraisedFromBegin = -10
+      reraisedFromEnd = -100
+
+when chronosStackTrace:
+  type StackTrace = string
+
+type
+  LocationKind* {.pure.} = enum
+    Create
+    Complete
+
+  CallbackFunc* = proc (arg: pointer) {.gcsafe, raises: [Defect].}
+
+  # TODO v4: consider removing this type and just use plain `CallbackFunc`
```

Comment:
This TODO could be removed because this task is almost impossible to achieve.

Reply:
what are current blockers?

Reply:
`udata` is get used everywhere

Reply:
they're actually not used all that frequently - at least not productively - we create a dynamic environment in many cases for other reasons so often whatever is in udata can be included in that environment instead - it's something I've thought about revisiting at some point but never had the time

Reply:
This `udata` is commonly used in Windows handling code, Future[T] combinators and in different waiting procedures. I wish to avoid create closures all the time i need to pass some integer/boolean value through callback.

Reply:
renamed to `InternalAsyncCallback` so as to avoid confusion about whether user code should use this type (it should not)